src/docs/rust.mdx

```diff
@@ -0,0 +1,324 @@
+---
+title: Rust Development at Sentry
+---
+
+This is a document that contains a bunch of useful resources for getting started with Rust and adhering to our coding principles.
+
+## Getting Started
+
+- Join the `#discuss-rust` channel in Slack
+- A quick introduction into the Syntax for first-timers:
+[https://fasterthanli.me/blog/2020/a-half-hour-to-learn-rust/](https://fasterthanli.me/blog/2020/a-half-hour-to-learn-rust/)
+- The Rust Book, comprehensively documenting the language:
+[https://doc.rust-lang.org/book/](https://doc.rust-lang.org/book/)
+- We follow the Rust style (enforced by `rustfmt` and `clippy` where possible):
+[https://doc.rust-lang.org/1.0.0/style/README.html](https://doc.rust-lang.org/1.0.0/style/README.html)
+- The Async Book:
+[https://rust-lang.github.io/async-book/](https://rust-lang.github.io/async-book/)
+
+## Advanced Reading
+
+- The Little Book of Rust Macros
+[https://danielkeep.github.io/tlborm/book/index.html](https://danielkeep.github.io/tlborm/book/index.html)
+- Rust Nomicon: The Dark Arts of Unsafe Rust
+[https://doc.rust-lang.org/nomicon/](https://doc.rust-lang.org/nomicon/)
+- Rust by Example: A comprehensive list of examples
+[https://doc.rust-lang.org/stable/rust-by-example/](https://doc.rust-lang.org/stable/rust-by-example/)
+- [https://doc.rust-lang.org/1.15.1/book/choosing-your-guarantees.html](https://doc.rust-lang.org/1.15.1/book/choosing-your-guarantees.html)
+
+
+## Coding Principles
+
+
+### Import order
+
+Imports must be declared before any other code in the file, and can only be preceded by module doc comments and module attributes (`#![...]`). We bundle imports in groups based on their origin, each separated by an empty line. The order of imports is:
+
+1. Rust standard library, including `std`, `core` and `alloc`
+2. External dependencies
+3. Workspace dependencies
+4. Crate modules
+5. Re-exports (`pub use`)
+
+**Example:**
+
+```rust
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::io::{self, Seek, Write};
+
+use fnv::{FnvHashMap, FnvHashSet};
+use num::FromPrimitive;
+
+use symbolic_common::{Arch, DebugId, Language};
+use symbolic_debuginfo::{DebugSession, Function, LineInfo};
+
+use crate::error::{SymCacheError, SymCacheErrorKind, ValueKind};
+use crate::format;
+
+pub use gimli::RunTimeEndian as Endian;
+```
+
+### Declaration order
+
+Within a file, the order of components should always be roughly the same:
+
+1. Module-level documentation
+2. Imports
+3. Public re-exports
+4. Modules and public modules
+5. Constants
+6. Error types
+7. All other functions and structs
+8. Unit tests
+
+Always declare a type or function before using it. The only exceptions to this are:
+
+- Public functions are always on top in `impl` blocks
+- Circular calls or references. In this case place the more significant item first. For example, for types that expose an iterator, declare the type and its impl block (including `fn iter`) first, then below define the corresponding iterator.
```

Comment:
Most folks seem to prefer to define high-level functions and types on top and helper functions / type internals further down, so as per a recent informal Slack poll about this, I suggest we remove this rule.

```suggestion
```

Reply:
I agree with this.

Reply:
I personally prefer public API on top. I don't know if it matters. I am not sure we can be prescriptive about this since I believe our codebase is all over the place already. The only thing we can say is that the tests module belongs at the bottom

Reply:
Agreed to start with important high level functions on top, and stuff you less frequently look at at the bottom.

I’m also a big fan of just splitting stuff up into smaller files. Not going too crazy like Java-style one file per class, but also not 1k+ line behemoths.

Reply:
> I’m also a big fan of just splitting stuff up into smaller files.

In theory, Rust module-level privacy already dictates what should be in a file: anything that's not supposed to access the private fields of a type should not be in the same file as that type. This would be my preference, but it does go in the direction of "one file per class", so I'm not sure how others think about it.